kernel/fde/optee/optee_impl.go

```diff
@@ -0,0 +1,342 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//go:build linux && (arm || arm64)
+
+/*
+ * Copyright (C) Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package optee
+
+/*
+#cgo CFLAGS: -I./
+#cgo LDFLAGS: -lteec
+#include <tee_client_api.h>
+#include <fde_key_handler_ta_type.h>
+#include <stdint.h>
+
+TEEC_UUID fde_ta_uuid = FDE_KEY_HANDLER_UUID_ID;
+
+TEEC_UUID pta_device_uuid = { 0x7011a688, 0xddde, 0x4053, \
+	{ \
+		0xa5, 0xa9, 0x7b, 0x3c, 0x4d, 0xdf, 0x13, 0xb8 \
+	} \
+};
+*/
+import "C"
+
+import (
+	"bytes"
+	"fmt"
+	"runtime"
+	"unsafe"
+
+	"github.com/snapcore/snapd/logger"
+)
+
+type opteeClient struct{}
+
+func (c *opteeClient) Present() bool {
+	pinner := &runtime.Pinner{}
+	defer pinner.Unpin()
+
+	// the first time we invoke the PTA without sending a buffer to fill. the
+	// PTA handles this by just returning the size of the buffer it will need.
+	bufferSize, err := devicesBufferSize(pinner)
+	if err != nil {
+		return false
+	}
+
+	// parameters:
+	// - output parameter containing a slice of bytes, each 16 byte segment is a UUID
+	// - none
+	// - none
+	// - none
+	params := teecParamTypes(C.TEEC_MEMREF_TEMP_OUTPUT, C.TEEC_NONE, C.TEEC_NONE, C.TEEC_NONE)
+	op := &C.TEEC_Operation{
+		started:    1,
+		paramTypes: params,
+	}
+	pinner.Pin(op)
+
+	output := make([]byte, bufferSize)
+	pinner.Pin(&output[0])
+
+	outputMemRef := unionAsType[C.TEEC_TempMemoryReference](&op.params[0])
+	outputMemRef.size = C.size_t(len(output))
+	outputMemRef.buffer = unsafe.Pointer(&output[0])
+
+	// this PTA (psuedo trusted application) command returns a list of PTA and
+	// early TA UUIDs. we know that our TA will always be an early TA.
+	if err := invoke(pinner, C.pta_device_uuid, 0x0 /* PTA_CMD_GET_DEVICES */, op); err != nil {
+		return false
+	}
+
+	// output here is a slice of bytes, each 16 byte segment is a UUID
+	output = output[:outputMemRef.size]
+
+	// fdeUUID is the byte encoding of the FDE TA UUID (C.fde_ta_uuid). Having this
+	// pre-calculated eliminates some code that would be required to do the
+	// conversion.
```

Comment:
Should it match `FDE_KEY_HANDLER_UUID_ID` in `kernel/fde/optee/fde_key_handler_ta_type.h`. Maybe you could add a reference here.